Added list read/write for symbols

RobertoRoos opened the pull request:
Fixes #266 

This adds the option to read multiple symbols at once, or write to multiple at once.

Unfortunately this is not directly compatible with `adsSumRead` and `adsSumWrite` (assuming we don't want the redundant info lookup when using only the name).
This is because those functions are based on the `SAdsSymbolInfo` structs, which contain the integer `dataType`, instead of a ctypes like `plc_type`.

I've tried to move some of the logic to separate functions so they can be reused, but the new symbol read/write methods resemble the classic functions a lot.

Example script:

```
from pyads import Connection
from random import random

plc = Connection('127.0.0.1.1.1', 851)

with plc:

    var_names = [
        "GVL.my_double",
        "GVL.another_double",
        "GVL.my_text",
        # "GVL.my_list"
    ]

    symbols = [plc.get_symbol(var) for var in var_names]

    new_data = [random(), random(), "Hello again"]

    data_with_names = dict(zip(var_names, new_data))

    # response = plc.write_list_by_name(data_with_names)
    # print(response)

    response = plc.write_list_symbols(symbols, new_data)
    print(response)

    # data = plc.read_list_by_name(var_names)
    # print(data)

    result = plc.read_list_symbols(symbols, True)
    print(result)

    for symbol in symbols:
        print(symbol.name + ":", symbol.value)

```

RobertoRoos committed:
Added auto-update feature (incl. doc update)

RobertoRoos committed:
Renamed auto_update method, made plc_type method static again

RobertoRoos committed:
Added kwarg to symbol init for auto_update

RobertoRoos committed:
Added auto_update keyword also to Connection.get_symbol

stlehmann committed:
Fix minor style and type-annotation issues from review

RobertoRoos committed:
Re-formated method in Connection

RobertoRoos committed:
Added destructor to pyads.Connection

RobertoRoos committed:
Added changelog entry and updated quickstart with mention of context manager and automatic close

stlehmann committed:
Merge pull request #179 from RobertoRoos/feature/connection

Destructor for pyads.Connection

RobertoRoos committed:
Set log levels back to DEBUG in testserver.py

stlehmann committed:
Merge branch 'master' into feature/ads_symbol

stlehmann committed:
Merge pull request #174 from RobertoRoos/feature/ads_symbol

Add AdsSymbol class

stlehmann committed:
Increase version to 3.3.1

stlehmann committed:
Update CHANGELOG.md

stlehmann committed:
Merge pull request #180 from stlehmann/stlehmann-patch-1

Increase version to 3.3.1

pylipp committed:
Fix some typos in documentation

stlehmann committed:
Version 3.3.2

pylipp committed:
Support structured variables in read_list_by_name

- optional argument for passing structure_defs into read_list_by_name
- store bytes returned by SUMUP_READ request
- convert to dict in read_list_by_name

stlehmann committed:
Merge pull request #182 from pylipp/fix-typos

Fix some typos in documentation

stlehmann committed:
Add value property

Automatically write value to plc if auto_update is True.

stlehmann committed:
Replace set_auto_update function by auto_update property.

Make the API more Pythonic.

stlehmann committed:
Update tests

stlehmann committed:
Update quickstart documentation

stlehmann committed:
Add test for writing if auto_update is True

stlehmann committed:
Update Changelog.md

stlehmann committed:
Use auto_update property

stlehmann committed:
Reformat

RobertoRoos committed:
Removed developing test-server (work-in-progress will be in a new branch)

RobertoRoos committed:
Updated setup.py to no longer require testserver_ex

stlehmann committed:
Merge pull request #192 from stlehmann/symbol

Make AdsSymbol even more pythonic

stlehmann committed:
Merge pull request #193 from RobertoRoos/bugfix/clean_testserver

Removed developing test-server

stlehmann committed:
Update CHANGELOG.md

stlehmann committed:
Merge pull request #194 from stlehmann/dev

Remove package testserver_ex

stlehmann committed:
Insert basic documentation chapters

stlehmann committed:
Replace symbols.md by symbols.rst

stlehmann committed:
Add routing.rst

stlehmann committed:
Update symbols.rst

stlehmann committed:
Update connection.rst

stlehmann committed:
Add testserver.rst

stlehmann committed:
Update device notification documentation

stlehmann committed:
Update Quickstart

stlehmann committed:
Merge branch 'documentation'

stlehmann committed:
Update Readme.md

stlehmann committed:
Update changelog

stlehmann committed:
Update version to 3.3.3

stlehmann committed:
Make plc_datatype for read_by_name optional

Also enable caching the symbol info for read_by_name

stlehmann committed:
Fix some small formatting and doc issues

stlehmann committed:
Allow write_by_name without plc_datatype

stlehmann committed:
Add tests for read/write without datatypes

stlehmann committed:
Fix tests

stlehmann committed:
Test non-caching

stlehmann committed:
Test for write_by_name to return None if connection is closed

stlehmann committed:
Update Changelog.md

stlehmann committed:
Update Changelog.md

stlehmann committed:
Merge pull request #195 from stlehmann/read_by_name_without_datatype

Read/write by name without datatype

stlehmann committed:
Update Readme.md

stlehmann committed:
Update Readme.md

stlehmann committed:
Update conda-forge link

stlehmann committed:
Fix some smaller style issues

stlehmann committed:
Rename port to value

stlehmann committed:
Fix some style issues

stlehmann committed:
Fix more style issues

stlehmann committed:
Fix some spelling issues

stlehmann committed:
Fix style issues on structs.py

stlehmann committed:
Fix style issues on testserver.py

stlehmann committed:
Update module headers, remove encoding statements

stlehmann committed:
Merge pull request #197 from stlehmann/reformat

Fix formatting, typing and spelling issues

chrisbeardy committed:
added int type hints to constants.py

stlehmann committed:
Merge pull request #199 from chrisbeardy/types_constants

added int type hints to constants.py

stlehmann committed:
Remove Python version checking.

We don't support Python 2 anymore.

stlehmann committed:
Merge remote-tracking branch 'origin/master'

stlehmann committed:
Fix link to Changelog.md in README.md

stlehmann committed:
Add conda-forge installation hint

stlehmann committed:
Fix broken pepy link in Readme.md

stlehmann committed:
Another Badge update

stlehmann committed:
Update documentation

Include datatype-querying for read_by_name and write_by_name.

Commit:
implementation, need tests

Commit:
implementation, need tests

Commit:
added tests for added max-sub-commands func

Commit:
added tests for added max-sub-commands func

Commit:
added new line at end of file

Commit:
added new line at end of file

stlehmann committed:
Fix read_state of AdvancedHandler

stlehmann committed:
Count variables

stlehmann committed:
Allow variable length for SUMP_READ

stlehmann committed:
Allow variable length for SUMP_WRITE

stlehmann committed:
Fix tests

stlehmann committed:
Support string datatype

stlehmann committed:
Restore test for SUM_WRITE with str

stlehmann committed:
Update Changelog

stlehmann committed:
Merge pull request #202 from stlehmann/fix_testserver

Testserver: SUMUP_READ and SUMUP_WRITE with variable length

stlehmann committed:
Fix string length and remove print-statement

stlehmann committed:
Alter tests to fit testserver quirks

stlehmann committed:
Merge remote-tracking branch 'chrisbeardy/split_read_write' into split_read_write

stlehmann committed:
Alter tests again after merge

stlehmann committed:
Update CHANGELOG.md

RobertoRoos committed:
Correct DT type, added support for DATE_AND_TIME and fixed AdsSymbol interpretation of strings

stlehmann committed:
Merge pull request #200 from chrisbeardy/split_read_write

Split read write by list into max-ads-sub-comands chunks

stlehmann committed:
Update CHANGELOG.md

stlehmann committed:
Merge branch 'master' into feature/types

stlehmann committed:
Merge pull request #206 from RobertoRoos/feature/types

Improved type handling by AdsSymbol

chrisbeardy committed:
quick additions of data type info

chrisbeardy committed:
removed duplicate tests and added addtional asserts to other read/wriet_list tests

removed some duplicate tests, possibly casuded by merge of last PR. Also addedmore test crtieria for the no cache tests to make sure that the symbol is not cached.

stlehmann committed:
Merge pull request #209 from chrisbeardy/more_tests

removed duplicate tests and added addtional asserts to existing tests

stlehmann committed:
Update CHANGELOG.md

stlehmann committed:
Merge pull request #208 from chrisbeardy/type_info

quick additions of data type info

pylipp committed:
Update Changelog

pylipp committed:
Support structured data types in write_list_by_name

- convert data values from dict into bytes, and write bytes
  accordingly in adsSumWrite()
- unit test still fails since data structure is lost during writing

stlehmann committed:
Fix docs

stlehmann committed:
fix tiny style issue

stlehmann committed:
Merge pull request #187 from pylipp/feature/structure-support-for-read-list-by-name

Support structured variables in read_list_by_name

stlehmann committed:
Add new-style type annotations

stlehmann committed:
Update Changelog.md

stlehmann committed:
Fix Union name conflict in ctypes and typing

RobertoRoos committed:
Added PLCTYPE_* as the prefered type of symbol initialization

stlehmann committed:
Merge pull request #212 from stlehmann/type-annotate-structs.py

Add new-style type annotations to structs.py

stlehmann committed:
Merge pull request #214

Added PLCTYPE_* as the prefered type of symbol initialization

stlehmann committed:
Add PLCSimpleDataType and PLCDataType

stlehmann committed:
Add PLCDataType to type annotations

stlehmann committed:
Rename symbol_type into data_type to comply with other functions

stlehmann committed:
Rename data_type to plc_datatype

stlehmann committed:
Merge branch 'master' into add-datatype-annotations

stlehmann committed:
Prevent Sphinx from unpacking PLCDataType

stlehmann committed:
Merge remote-tracking branch 'origin/add-datatype-annotations' into add-datatype-annotations

stlehmann committed:
Fix get_symbol docstring

stlehmann committed:
Merge pull request #215 from stlehmann/add-datatype-annotations

Add datatype annotations

stlehmann committed:
Change type comments to annotations

stlehmann committed:
Remove test.py

stlehmann committed:
Fix handle_count by using class variable instead of instance variable

stlehmann committed:
Fix tests in test_connection_class.py

stlehmann committed:
Force to provide all PLCVariable properties on initialization

We do that to make variables more determined and less magical. We sacrifice a little convenience, though.

stlehmann committed:
Do not allow variable creation on write

This makes variable creation a little less convenient but leads to clearer test documentation as all variables must be declared before one can read or write to it.

stlehmann committed:
Pack value if basic datatype

stlehmann committed:
Fix tests in test_connection_class.py

stlehmann committed:
Fix tests in test_symbol.py

stlehmann committed:
Merge pull request #216 from stlehmann/elaborate_advanced_handler

Make the AdvancedHandler less magical

stlehmann committed:
Create empty dict if structure_defs is None

stlehmann committed:
Make structured_data_names optional

stlehmann committed:
Use AdvancedHandler for test function

stlehmann committed:
Add dummy condition for sumup_write

Needs to be filled with life

stlehmann committed:
Implement sumup_write on testserver.AdvancedHandler

stlehmann committed:
Make structured_data_names optional in adsSumRead

stlehmann committed:
Add sumup_read to testserver.AdvancedHandler

stlehmann committed:
Fix error caused by modified list parameter

stlehmann committed:
Fix tests

stlehmann committed:
Update Changelog.md

stlehmann committed:
Make structured_data_names non-optional

The use-case where no structure_def is passed is covered by the ads.write_list_by_name and read_list_by_name functions.

stlehmann committed:
Merge pull request #220 from stlehmann/structure_support_for_write_list_by_name

Structure support for write list by name

stlehmann committed:
Version 3.3.4

stlehmann committed:
Create python-publish.yml

Auto-upload package to PyPi

stlehmann committed:
Create ci.yml

First test of Github Actions

stlehmann committed:
Fix dependency install

stlehmann committed:
Try setup.py develop

stlehmann committed:
Run make

stlehmann committed:
Checkout submodule

stlehmann committed:
Compile adslib

stlehmann committed:
Fix compile command

stlehmann committed:
Use setup.py

stlehmann committed:
Check coverage

stlehmann committed:
Rename workflos to CI

stlehmann committed:
Update badge in README.md

stlehmann committed:
Delete .travis.yml

stlehmann committed:
Update CHANGELOG.md

stlehmann committed:
Merge pull request #221 from stlehmann/stlehmann-patch-1

Use Github Actions for CI

stlehmann committed:
Fix docs for adding routes to target

Add warning to open port before setting local netid.

stlehmann committed:
Fix index.rst

stlehmann committed:
Fix ads.py

stlehmann committed:
Fix constants.py

stlehmann committed:
Fix structs.py

stlehmann committed:
Fix pyads.rst

stlehmann committed:
Fix error int test_correct_route

stlehmann committed:
Merge pull request #222 from stlehmann/fix-docs

Fix  some documentation issues

stlehmann committed:
Add structure_def parameter

stlehmann committed:
Simplify send_raw_udp_message

stlehmann committed:
Fix mypy errors in pyads_ex.py

stlehmann committed:
Remove array_size parameeter from bytes_from_dict

This is done because the array size is defined by the length of the list.

stlehmann committed:
Fix ads.py

stlehmann committed:
Rename _read_write_check to _check_for_open_connection

Just to make the function more descriptive

stlehmann committed:
Remove check for self._plc

The connection is a non-optional parameter

stlehmann committed:
Add return type annotations

stlehmann committed:
Allow redefinitions

stlehmann committed:
Support structure for read and write

stlehmann committed:
Fix type annotations

stlehmann committed:
mypy: don't use strict_optional

Because this is so annoying

RobertoRoos committed:
Added proper capture of char arrays inside pyads_ex

RobertoRoos committed:
Adder chararray capture in write operation too, added string tests for AdsSymbol

RobertoRoos committed:
Dug deeper and replaced more string comparisons

chrisbeardy committed:
updated notification docs to use the parse_notification methods

stlehmann committed:
Fix style issues

stlehmann committed:
Add structure_def parameter to Connection.get_symbol

stlehmann committed:
Add tests

stlehmann committed:
Add __future__.annotations

To prevent Sphinx from expanding custom types.

stlehmann committed:
Add structure_def parameter to docstring

stlehmann committed:
Let sphinx auto-generate __init__ docstrings

stlehmann committed:
Prevent type expansion in docs

stlehmann committed:
Add structure_def parameter to docs

stlehmann committed:
Add array_size parameter

stlehmann committed:
Multiply structure_size by array_size

stlehmann committed:
Add tests for arrays

stlehmann committed:
Add string to tests

stlehmann committed:
Add changelog entry

stlehmann committed:
Modify docs

stlehmann committed:
Merge pull request #226 from chrisbeardy/update_notification_docs

updated notification docs to use the parse_notification methods

stlehmann committed:
Merge pull request #225 from RobertoRoos/feature/strings

Added proper capture of char arrays

stlehmann committed:
Merge pull request #223 from stlehmann/add_struct_support_to_AdsSymbol

Add structure support to AdsSymbol class

RobertoRoos committed:
Refactored testserver into multiple files

chrisbeardy committed:
added missing LINt type from datatype_map

stlehmann committed:
Upgrade requirements

stlehmann committed:
Merge pull request #242 from stlehmann/update-requirements

Upgrade requirements

stlehmann committed:
Merge pull request #238 from chrisbeardy/patch-1

Add missing LINT type from datatype_map

stlehmann committed:
Update CHANGELOG.md

stlehmann committed:
Replace PORT_SPS1 by PORT_TC3PLC1 in documentation.

For TwinCAT3 the default port changed from 801 to 851. As TwinCAT3 is out there for a while now the documentation now reflects the new default port.

stlehmann committed:
Remove omit testserver

stlehmann committed:
Remove shebang

stlehmann committed:
Add testserver tests

stlehmann committed:
Fix CI

stlehmann committed:
CI: Indicate completion of coveralls.io

stlehmann committed:
Add tests for AdvancedHandler

stlehmann committed:
Update CHANGELOG.md

stlehmann committed:
Merge pull request #243 from stlehmann/testserver-package

Refactor testserver into multiple files

RobertoRoos committed:
Add support for testserver notifications (#239)

stlehmann committed:
Increase version to 3.3.5

stlehmann committed:
CI: Upload Python Package manually

stlehmann committed:
CI: Upload only source distributions

stlehmann committed:
Drop support for Python < 3.7

stlehmann committed:
Deploy: sync submodule

RobertoRoos committed:
Add pyads.testserver to setup.py (#249)

stlehmann committed:
Version 3.3.6

stlehmann committed:
Deploy: Fix submodules

stlehmann committed:
Version 3.3.7

stlehmann committed:
Fix initial length of read data (#264)

Remove c_ulong and use 4 Bytes instead because depending on arcitecture c_ulong can be 4 or 8 bytes.

stlehmann committed:
Update CHANGELOG.md

stlehmann committed:
Increase version number

dependabot[bot] committed:
Bump urllib3 from 1.26.4 to 1.26.5 (#250)

Bumps [urllib3](https://github.com/urllib3/urllib3) from 1.26.4 to 1.26.5.
- [Release notes](https://github.com/urllib3/urllib3/releases)
- [Changelog](https://github.com/urllib3/urllib3/blob/main/CHANGES.rst)
- [Commits](https://github.com/urllib3/urllib3/compare/1.26.4...1.26.5)

---
updated-dependencies:
- dependency-name: urllib3
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

RobertoRoos committed:
Added read/write by list feature for symbols (this branch was rebased on top of the refactor)

RobertoRoos commented:
@stlehmann I was hoping you might already have some insights here.
Would you like less overlap with the old `*_list_by_name` functions?
Do the new interfaces seem right?

chrisbeardy commented:
This looks like a pretty good additional feature, nice work! The symbol way of using pyads I think is fast becoming a favourite from the looks of things. I can already see a use case to optimise some of my code to use the reading a list of symbols.

I haven't looked through the implementation in detail yet, but some fast thoughts with my opinion:

- My thought is this is going to be a well used feature, as I said as soon as I noticed this PR and your example I got excited
  - given this addition there will sort of be two competing ways to use pyads from an end users perpsective so time should be given to make sure this is straight forward and clear to the end user what all methods do but also maintain familarity between them.
  - Also given my point that I think it will be well used, time should be given to make sure it's straightforward and clear for the maintainer and contributors of pyads to maintain.
- I believe changing the method names to `read_list_of_symbols` and `write_list_of_symbols` may read better to the user, gives an accurate description of what the method is doing and isn't that much longer for a name.
- Possibly could think about the parameters for the `write_list_symbols`; would it make sense to make this a dictionary so the interface is in line with `write_list_by_name` with the keys being the symbol and the values the value to write to that symbol. There is an argument that could be had to which one is a nicer interface (two lists or a dict) and which one is optimal, but since `write_list_by_name` and to some extend `write_structure_by_name` use dictionaries as input I believe consistency across the module would be beneficial.
- Have `read_list_symbols` always return the dictionary and do away with the optional parameter. Again for consistency with `read_list_by_name` and also I believe the user will likely be expecting something back when asking to read something from the plc even though in this case it is not actually required. I expect the time overhead of additionally inserting the keys into a dictionary will not be of importance? Alternatively just swap the optional parameter around so the default is to return.
- I am wondering if it is possible to modify `adsSumReadBytes` to have the same interface as `adsSumRead` but instead just return bytes instead of doing the processing as you say. I'm not sure it is possible to modify `data_symbols: List[Tuple[int, int, int]]` into  `data_names: List[str], data_symbols: Dict[str, SAdsSymbolEntry],
    structured_data_names: List[str]` and then that processing for the conversion can be pushed into the method that calls `adsReadSumBytes`. 
  - The singledispatch library could be considered (or two methods or using more private methods with more descriptive names or another solution) so that am `adsSumReadBytes` like method could exist for datanames, then we can have `adsReadSum` just call that. Also down the same lines, just have an `adsReadSum` which accepts both options. 
  - I think my main concern here is that as you say there is an overlap with the existing functions, but I think that is ok and can not be avoided to some extent as both methods (by dataname and by symbol) are both just reading bytes from the PLC and converting them into the desired format at the end of the day. I think it just needs to be thought about and made clear to why there are similar methods that appear to do a similar thing but have different parameters.
  - From an end users perspective this is just the internals of the library but I think for the maintenance of the library going forward it is important. 
  - Is there scope to potentially have `read_list_of_symbols` just wrap around `read_list_by_name` with the addition of updating the symbol cache. As the symbol contains the dataname. I guess this becomes memory inefficient as you then build up two caches and adds an extra overhead of having to fetch the handle twice.
  - I am sure @stlehmann will have a better understanding of the organisation of the library and be able to give more thoughts
- I am not 100% sure if it is applicable here , but I know beckhoff advise that no more that 500 symbols/requests are read/written in one ADS call. Therefore in the implementation of `read/write_list_by_name` the list of datanames is chunked up into number of `ads_sub_commands` blocks. I think this is required here too as there is still a number of ads requests:
![image](https://user-images.githubusercontent.com/20585410/135508884-794f77ad-c169-449c-860d-7f68e6068b47.png)
- Could consider renaming `convert_data_to_value` method to again be a more specific and descriptive (naming is always hard and contentious so I apologise if this seems petty). possibly something like `get_value_from_ctype_data`.
- Could also consider refactoring to be more like `read_list_by_name` where the conversion from bytes to relevant symbol data function is in `pyads_ex.py`, then can make `convert_data_to_value` a private method that doesnt span modules, possibly making `ads.py` easier to read. Again I'm sure @stlehmann will have a better thought on this. 
- It could be worth splitting this work into two PRs, purely from an admin perspective, one that refactors `pyads_ex.py` and adds in the `adsReadSumBytes`, potentially making it compatible with the existing `adsReadSum`. Then another PR for actually adding in the `read/write_list_symbols` feature. This way I think it may make it easier to test and make sure the first PR and refactoring of the `pyads_ex.py` file does not break anything given the difficulty in thoroughy testing this library. This could be made into a minor release, giving maintainers a chance to easily debug in the future. 
- Likely you were planning on it but tests and docs wil need to be added too.

My apolgies for what grew into a longer list than I thought as I spent more time thinking about it but I hope that is food for thought and helps you and @stlehmann guide this feature smootly into the library.

Also, I know that I am not a maintainer of pyads, but I have contributed and I use it a lot so I have an interest in the features that get added so thank you again for your work on the AdsSymbol side of things!

RobertoRoos commented:
Whoof, a lot. But thanks @chrisbeardy , that's very useful.

> I believe changing the method names to read_list_of_symbols and write_list_of_symbols may read better to the user, gives an accurate description of what the method is doing and isn't that much longer for a name.

Agreed, I'll change that.

> Possibly could think about the parameters for the write_list_symbols; would it make sense to make this a dictionary so the interface is in line with write_list_by_name with the keys being the symbol and the values the value to write to that symbol. There is an argument that could be had to which one is a nicer interface (two lists or a dict) and which one is optimal, but since write_list_by_name and to some extend write_structure_by_name use dictionaries as input I believe consistency across the module would be beneficial.

Yep, was thinking about that too. I'll change it.
Though nothing stops us from accepting both? We could detect if the first argument is a list or dict, and take it from there? Though that might make the interface too complicated.

> Have read_list_symbols always return the dictionary and do away with the optional parameter. Again for consistency with read_list_by_name and also I believe the user will likely be expecting something back when asking to read something from the plc even though in this case it is not actually required. I expect the time overhead of additionally inserting the keys into a dictionary will not be of importance? Alternatively just swap the optional parameter around so the default is to return.

Also agreed. Will change.

> Could consider renaming convert_data_to_value method to again be a more specific and descriptive (naming is always hard and contentious so I apologise if this seems petty). possibly something like get_value_from_ctype_data.

Will change.
No, I don't find it nick-picky! This part of writing good code, and in terms of naming I can still learn some.

Alright, I think that most of the concrete points. I guess now is the part where we debate about structure and details.

About the refactoring stuff, that is a good point. Though I think we need both interfaces (`read_list_by_name` and `read_list_of_symbols`) in order to write a good shared low-level base.
I'm personally not too worried about things breaking. The changes I made so far are no big changes. But I understand wanting caution.

RobertoRoos commented:
I've started a _refactor_ branch containing the refactors I did in this branch (but without the new interface): https://github.com/RobertoRoos/pyads/tree/feature/refactor

We can create a pull request for that, and retarget this PR into that one to split the diffs.

chrisbeardy commented:
I like this as an API for the user.

I agree that it may be the case the user is using symbols and lists and then it would be irritating to force them to convert this to a dictionary just for the input to the write method. My only slight concern is the use of advising the user to write to a private value of the symbol class. I guess the "proper" way do do it would be to to make sure `auto_update` is off and then write the property `value` instead but I can understand why that would be awkward and confusing to a user who doesn't care about the implementation. I guess as long as it is properly documented then I see no big harm. I am also happy with the writing of the private variables in the implementation of `read/write_list_of_symbols` as they are tightly coupled to the symbol class anyway and can be dealt with using unit tests. 

The only other change to be considered from a user API perspictive is the `ads_sub_commands` option which by default should be set to 500.

As you say that I believe that is the concrete / user perspective points. The rest is internal implementation, this can be discussed and I believe @stlehmann is likely best placed to have a view on this when they have time available. Although as it is internal and I agree your changes are not huge, there is always scope to just refactor the internals at a later date if/when required. 

As for naming, I discovered this [cheatsheet](https://github.com/kettanaito/naming-cheatsheet/blob/master/README.md) a little while ago and I thing it sums up and makes some good suggestions.

chrisbeardy commented:
Also I forgot to mention, i like the use of

```
Comparable to :meth:`Connection.read_list_by_name`.
See also :class:`pyads.AdsSymbol`.
```

in the docstrings.

stlehmann commented:
Sorry to keep you guys waiting. I'm a bit busy at work at the moment and hardly finding any sparetime for development. @RobertoRoss I'll try to give it a closer look during the week.

RobertoRoos committed:
Moved Connection to its own file, did some refactoring in ads files (#269)

stlehmann committed:
Add docs for connection module

stlehmann committed:
Add changelog entry

kryskool committed:
Add constants for ports TC3PLC2 to TC3PLC4 (#272)

* Adds constant for ports TC3PLC2 to TC3PLC4

LIke TC2, TC3 have 4 primary ports equivalent.
see https://infosys.beckhoff.com/english.php?content=../content/1033/tcplclib_tc2_system/31064331.html&id=

* Update CHANGELOG.md

na-svendlich committed:
Enable pyads to use with tc/bsd (#247)

* Enable pyads to use in new tc/bsd

Just a small change to enable pyads to be used under tc/bsd: https://www.beckhoff.com/de-de/produkte/neuheiten/twincat-bsd/

* Create new function platform_is_freebsd

* Use function platform_is_freebsd to identify bsd os

Co-authored-by: Stefan Lehmann <stlm@posteo.de>

stlehmann committed:
Update Changelog

Commit:
Add hint on unidirectional routes in docs (#258)

* More explicitely explain route settings

fixes #255

* Change note

Co-authored-by: Stefan Lehmann <stlm@posteo.de>

RobertoRoos commented:
@stlehmann yeah I never rebased this branch onto the other refactoring branch. I'll do that next.

About the double interfaces, I don't think splitting this package would be necessary. I would say all real functionality remains in separate functions, and the Symbol class just references those function.
I find it hard to estimate how confusing the two methods are to new Pyads users. I think it's manageable.

stlehmann commented:
Yes, a rebase would be good. You can rebase on master because the refactoring branch has already been merged.

stlehmann committed:
Update requirements

stlehmann committed:
Add tests for Python 3.9

stlehmann committed:
Remove Python 3.6 from tox

RobertoRoos committed:
Started with new unit tests for list feature

stlehmann commented:
Ah this looks much better now. Thanks for rebasing. 👍

RobertoRoos committed:
Split code into two levels inside Connection and made pyads_ex list functions compatible with tuples

RobertoRoos committed:
Moved list name retrieval to own method + ran black on three files

stlehmann reviewed (changes requested):
@RobertoRoos please remove the reformatting from the PR. We had this a few times now and I won't accept any PRs refactoring or reformatting other parts of the code anymore. If you see something that needs reformatting please raise an issue or make a separate PR for this.

RobertoRoos committed:
Undid black reformatting on old code

stlehmann commented:
> 
> 
> Okay, I've created the above. I think this could work, I like the unit-purposes:
> 
>     * `read_list_by_name` handes the name-info retrieval while `read_list_of_symbols` only processes the symbol objects
> 
>     * `_read_list` queries the pyads_ex back end to get Python values out of symbol infos (also handles ads query limit)
> 
>     * `adsSumRead` gets binary data from symbol info and turns that into Python values (regardless of the type of info storage)
> 
> 
> @stlehmann , thoughts?

I gave this some thoughts and I come to dislike the idea of different functions for symbols and names more and more. The Symbols approach should integrate smoothly in the current API and shouldn't feel so much like a complete different way of doing things with pyads. So my preferred way is to with Connection.read_list and Connection.write_list processing both strings and symbols and given some time deprecate the read_list_by_name function.

RobertoRoos commented:
Alright, that's understandable. I'll modify the code (but I think I'll archive the current version on my fork just in case).

I intend to approach it with an `if ... else ...` in `read_list()`, calling one of two private methods to do the real read.

stlehmann commented:
Yes, I think a simple if ... else should do to separate the two cases.

RobertoRoos committed:
Added combined _list interface methods for names and symbols

RobertoRoos commented:
And there it is. I started with updating the docs too.

stlehmann reviewed (commented):
@RobertoRoos Thank you very much for your many efforts on this issue. 

Sadly I must tell you that I'm more than a bit reluctant to apply your changes because they add so much complexity and are hard to overview and will be even harder to maintain. You must understand that pyads is just a small side-project of mine and I have limited time for it. So my priority is to keep the package up-to-date and maintain the code-base. Additional features are appreciated but should be implemented in a straight-forward way with minimal changes to the existing codebase. This PR applies many fundamental changes to the existing codebase. Plainly speaking it is just too big for me to overview. And even though I'd like to have this feature implemented I think it is not worth all the refactoring, at least at the moment. I'm more than willing to keep this PR open and check if we can implement it at some later point in time. Also you are more than welcome to work on the codebase in small steps to make the necessary changes to apply this feature smoothly in the future.

Kind Regards
Stefan

RobertoRoos commented:
@stlehmann , alright, that makes sense, no worries.
I think I'll just merge this into my own project, so it's the same to me in the end.

chrisbeardy commented:
In the meantime @RobertoRoos  you could look at making a separate pypi package as an extension of pyads, called pyads-symbols as Stefan suggested. This would allow you to work downstream of pyads and add lots of symbol features. This could make pyads easier to maintain in the future and give the symbols some more freedom. It can then always be merged back in later if deemed wise at the time.

stlehmann commented:
> In the meantime @RobertoRoos you could look at making a separate pypi package as an extension of pyads, called pyads-symbols

I very much support this idea. It has many benefits in terms of maintainability and it makes it easier to implement new features on top of pyads without bloating the pyads package too much. A first step could be to move the symbols module there and then go on and add new features successively
